File: leetcode/algorithms/20_valid_parentheses/main.cpp
#include <stack>
#include <string>
#include <unordered_map>
using namespace std;

class ValidParentheses {
public:
    /**
     * LIFO (Last-In, First-Out): Stack
     * 
     * Complexities:
     *   N - Length of `s`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(N)
     */
    bool isValid(string s) {
        stack<char> char_stack;

        for (char c : s) {
            switch (c) {
                case '(':
                    char_stack.push(')');
                    break;
                case '[':
                    char_stack.push(']');
                    break;
                case '{':
                    char_stack.push('}');
                    break;
                default:
                    if (char_stack.empty() || char_stack.top() != c) {
                        return false;
                    }
                    char_stack.pop();
                    break;
            }
        }

        return char_stack.empty();
    }


    // Solution
    /**
     * LIFO (Last-In, First-Out): Stack
     * 
     * Complexities:
     *   N - Length of `s`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(N)
     */
    bool solution(string s) {
        stack<char> st;

        for (char c : s) {
            if (c == '(') {
                st.push(')');
            } else if (c == '{') {
                st.push('}');
            } else if (c == '[') {
                st.push(']');
            } else {
                if (st.empty() || st.top() != c) {
                    return false;
                }
                st.pop();
            }
        }

        return st.empty();
    }
};

File: leetcode/algorithms/20_valid_parentheses/main.java
import java.util.Stack;

class ValidParentheses {
  /**
   * LIFO (Last-In, First-Out): Stack
   * 
   * Complexities:
   *   N - Length of `s`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(N)
   */
  public boolean isValid(String s) {
    Stack<String> stack = new Stack<>();

    for (String c : s.split("")) {
      switch (c) {
        case "(" -> stack.push(")");
        case "[" -> stack.push("]");
        case "{" -> stack.push("}");
        default -> {
          if (stack.empty() || !stack.pop().equals(c)) {
            return false;
          }
        }
      }
    }

    return stack.empty();
  }


  // Solution
  /**
   * LIFO (Last-In, First-Out): Stack
   * 
   * Complexities:
   *   N - Length of `s`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(N)
   */
  public boolean solution(String s) {
    if (s.length() % 2 != 0) {
      return false;
    }

    Stack<Character> stack = new Stack<>();

    for (char c : s.toCharArray()) {
      switch (c) {
        case '(' -> stack.push(')');
        case '{' -> stack.push('}');
        case '[' -> stack.push(']');
        default -> {
          if (stack.isEmpty() || stack.pop() != c) {
            return false;
          }
        }
      }
    }

    return stack.isEmpty();
  }
}
